src/Statistics.jl

```diff
@@ -0,0 +1,286 @@
+the starting point of the algorithm.
+"""
+function mean!(M::Manifold, y, x::AbstractVector, w::AbstractWeights = (n = length(x); Weights(ones(n), n));
+            stop_iter=100,
+            kwargs...
+        ) where {T}
+    yold = similar_result(M, mean, y)
+    copyto!(yold,y)
+    v0 = zero_tangent_vector(M, y)
+    v = map(_ -> copy(v0), x)
+    wreg = convert(Vector, w) ./ (2 * w.sum)
+    for i=1:stop_iter
+        copyto!(yold,y)
+        log!.(Ref(M), v, Ref(yold), x)
+        vreg = sum(wreg .* v)
+        exp!(M, y, yold, vreg)
+        isapprox(M,y,yold; kwargs...) && break
+    end
+    return y
+end
+
+@doc doc"""
+    median(M, x, weights=Weights(ones(n)); stop_iter=10000)
+
+computes the Riemannian median of the vector `x`  of `n` points on the
+[`Manifold`](@ref) `M`. This function is nonsmooth (i.e nondifferentiable) and
+uses a cyclic proximal point scheme. Optionally one can provide
+weights $w_i$ for the weighted Riemannian median. The general formula to compute
+the minimizer reads
+````math
+\argmin_{y\in\mathcal M}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemannian [`distance`](@ref).
+
+Optionally you can provide `x0`, the starting point (by default set to the first
+data point). `stop_iter` denotes the maximal number of iterations to perform and
+the `kwargs...` are passed to [`isapprox`](@ref) to stop, when the minimal change
+between two iterates is small. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in Algorithm 4.3 and 4.4 in
+> Bačák, M: Computing Medians and Means in Hadamard Spaces.
+> SIAM Journal on Optimization (2014), 24(3), pp. 1542–1566,
+> doi: [10.1137/140953393](https://doi.org/10.1137/140953393),
+> arxiv: [1210.2145](https://arxiv.org/abs/1210.2145)
+"""
+function median(M::Manifold, x::AbstractVector, w::AbstractWeights = (n = length(x); Weights(ones(n), n)); x0=x[1], kwargs...)
+    y = similar_result(M, median, x[1])
+    copyto!(y,x0)
+    return median!(M, y, x, w; kwargs...)
+end
+@doc doc"""
+    median!(M,y,x,w)
+
+computes the [`median`](@ref) in-place in `y` where the initial value of `y` is
+the starting point of the algorithm.
+"""
+function median!(M::Manifold, y, x::AbstractVector,
+    w::AbstractWeights = (n = length(x); Weights(ones(n), n));
+    stop_iter=1000000,
+    kwargs...
+) where {T}
+    n = length(x)
+    yold = similar_result(M,median,y)
+    copyto!(yold,y)
+    (length(w) != n) && throw(DimensionMismatch("The number of weights ($(length(w))) does not match the number of points for the median ($(n))."))
+    v = zero_tangent_vector(M,y)
+    wv = convert(Vector, w) ./ w.sum
+    for i=1:stop_iter
+        λ =  .5 / i
+        copyto!(yold,y)
+        for j in 1:n
+            @inbounds t = min( λ * wv[j] / distance(M,y,x[j]), 1. )
+            @inbounds log!(M, v, y, x[j])
+            y = exp(M, y, v, t)
```

Comment:
You know, we should probably use mutating `exp!` here

Reply:
No, we discussed that already, since `y`is already the base point it should not be mutable, that might lead to unwanted side effects.

Reply:
But we can introduce a `ytmp` allocated outside the loop that is mutated here and then copied to `y` on the next line. That would replace the allocation with a copy. I have an implementation that passes the tests if you don't mind me pushing.

Reply:
If you think its worth the effort, yes, why not. I like the current version for its readability.